controlplane/kubeadm/controllers/controller.go

```diff
@@ -255,11 +268,23 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 	}
 
 	controlPlane := internal.NewControlPlane(cluster, kcp, ownedMachines)
-	requireUpgrade := controlPlane.MachinesNeedingUpgrade()
+
+	// Aggregate the operational state of all the machines; while aggregating we are adding the
+	// source ref to the aggregate reason (reason@machine/name) so the problem can be easily tracked down to its source.
+	conditions.SetAggregateCondition(controlPlane.KCP, controlplanev1.MachinesReadyCondition, ownedMachines.ConditionGetters(), conditions.AddSourceRef())
```

Comment:
From an external consumer point of view, would it make sense to try and differentiate between degraded and broken here somehow? Currently we leave that up to the user, but it seems like we should have enough context within KCP to differentiate between the two (especially since any upgrade or scaling operations would cause this to become False)

Reply:
I think this makes sense, but IMO those are separated Conditions